kentik_api_library/kentik_api/analytics/data_frame_cache.py

```diff
@@ -0,0 +1,256 @@
+        # We have to handle this case to make collaboration easier
+        if name.count("_") == 1:
+            # happy default case, 1 underscore => no mangling
+            try:
+                s, e = name.split(cls.separator)
+            except ValueError as exc:
+                log.critical("Invalid filename: %s (e: %s)", filename.name, exc)
+                return None, None
+            try:
+                start = datetime.fromisoformat(s)
+                end = datetime.fromisoformat(e)
+                return start, end
+            except ValueError as exc:
+                log.critical("Invalid filename component %s", exc)
+                return None, None
+        else:
+            # Assume that all colons in the original name which is composed of 2 ISO datetime strings with timezone
+            # offset were replaced by underscores
+            # ISO format datetime with timezone offset is always exactly 25 characters
+            # So, the name must be exactly 51 characters to contain 2 valid ISO datetime strings separated by '_'
+            # and the character in the middle (index 25) must be '_'
+            log.debug("Got mangled name: %s", filename)
+            if len(name) != 51 or name[25] != cls.separator:
+                log.critical("Invalid filename: %s", filename)
+                return None, None
+            try:
+                start = datetime.fromisoformat(name[:25].replace("_", ":"))
+                end = datetime.fromisoformat(name[26:].replace("_", ":"))
+                return start, end
+            except ValueError as exc:
+                log.critical("Invalid (mangled) filename component %s", exc)
+                return None, None
+
+    def __init__(self, directory: Path) -> None:
+        self.data_dir = directory
+        if not self.data_dir.exists():
+            self.data_dir.mkdir()
+        else:
+            if not directory.is_dir():
+                raise RuntimeError("{}: Path is not a directory: {}".format(self.__class__, directory))
+        log.debug("data_dir: %s", self.data_dir)
+        log.debug("oldest: %s newest: %s", self.oldest, self.newest)
+
+    def __repr__(self) -> str:
+        return f"DFCache: dir: {self.data_dir.resolve()}"
+
+    @property
+    def files(self) -> Generator[Path, None, None]:
+        for f in sorted(self.data_dir.glob(self.filename_format.format(start="*", end="*"))):
+            yield f
+
+    @property
+    def file_count(self) -> int:
+        return len(list(self.files))
+
+    @property
+    def data_size(self) -> int:
+        return sum(f.stat().st_size for f in self.files)
+
+    @property
+    def is_empty(self) -> bool:
+        return self.file_count == 0
+
+    @property
+    def oldest(self) -> Optional[datetime]:
+        match = self.filename_format.format(start="*", end="*")
+        try:
+            first = sorted(list(self.data_dir.glob(match)))[0]
```

Comment:
Not sure here: does the result, sorted alphabetically, is equivalent to result sorted by datetime encoded in the filenames?

Reply:
Yes, files are named with `<start-time>_<end-time>` with both times in ISO format. Alphabetic sorting provides sorting according to start time and for equal start times by end time.

Reply:
OK, just wanted to make sure